packages/runtime-core/src/renderer.ts

```diff
@@ -1535,6 +1535,11 @@ function baseCreateRenderer(
           popWarningContext()
         }
       }
+
+      // #6930 avoid infinite recursion when reloading hmr components
+      if (__DEV__ && hmrDirtyComponents.has(instance.type)) {
+        hmrDirtyComponents.delete(instance.type)
```

Comment:
Good catch!
I feel we should delete the dirty component here: https://github.com/vuejs/core/blob/b0b74a160c08941fa9a7a5460f36a1f2fccbf423/packages/runtime-core/src/hmr.ts#L138

change the code as blow:
```javascript
queueJob(() => {
  instance.parent.update()
  hmrDirtyComponents.delete(oldComp)
})
```

Reply:
Thanks a lot for the suggestion!